nav2_rviz_plugins/src/nav2_panel.cpp

```diff
@@ -93,6 +173,108 @@ Nav2Panel::onShutdown()
   QFuture<void> future =
     QtConcurrent::run(std::bind(&nav2_lifecycle_manager::LifecycleManagerClient::shutdown,
       &client_));
+  timer_->cancel();
+}
+
+void
+Nav2Panel::onCancel()
+{
+  QFuture<void> future =
+    QtConcurrent::run(std::bind(&Nav2Panel::onCancelButtonPressed,
+      this));
+}
+
+void
+Nav2Panel::onNewGoal(double x, double y, double theta, QString frame)
+{
+  auto pose = geometry_msgs::msg::PoseStamped();
+
+  pose.header.stamp = rclcpp::Clock().now();
+  pose.header.frame_id = frame.toStdString();
+  pose.pose.position.x = x;
+  pose.pose.position.y = y;
+  pose.pose.position.z = 0.0;
+  pose.pose.orientation = orientationAroundZAxis(theta);
+
+  startNavigation(pose);
+}
+
+
+geometry_msgs::msg::Quaternion
+Nav2Panel::orientationAroundZAxis(double angle)
+{
+  tf2::Quaternion q;
+  q.setRPY(0, 0, angle);  // void returning function
+  return tf2::toMsg(q);
+}
+
+void
+Nav2Panel::onCancelButtonPressed()
+{
+  auto future_cancel = action_client_->async_cancel_goal(goal_handle_);
+
+  if (rclcpp::spin_until_future_complete(client_node_, future_cancel) !=
+    rclcpp::executor::FutureReturnCode::SUCCESS)
+  {
+    RCLCPP_ERROR(client_node_->get_logger(), "Failed to cancel goal");
+    return;
+  }
+}
+
+void
+Nav2Panel::timerActionEvent()
+{
+  if (!goal_handle_) {
+    RCLCPP_DEBUG(client_node_->get_logger(), "Waiting for Goal");
+    state_machine_.postEvent(new ROSActionQEvent(false));
+    return;
+  }
+
+  rclcpp::spin_some(client_node_);
+  auto status = goal_handle_->get_status();
+
+  // Check if the goal is still executing
+  if (status == action_msgs::msg::GoalStatus::STATUS_ACCEPTED ||
+    status == action_msgs::msg::GoalStatus::STATUS_EXECUTING)
+  {
+    state_machine_.postEvent(new ROSActionQEvent(true));
```

Comment:
If I understand this correctly, you are posting a bool as an event. Is that correct? That conveys little information. The events should be an enum at a minimum. That way you could post something like "NAV_ACTIVATED"  or "NAV_DEACTIVATED"

Reply:
Same functionality but I suppose an enum indicates it is more specific to actions being active/inactive.